### PR TITLE
feat: Log FTT2 loss + Fourier Correlation loss 

### PR DESCRIPTION
**Pull request by OpheliaMiralles**
## Type of Change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update

## Issue Number

new loss function

## Code Compatibility

-   [x] I have performed a self-review of my code

### Code Performance and Testing

-   [ ] I have added tests that prove my fix is effective or that my feature works
-   [x] I ran the [complete Pytest test](https://anemoi.readthedocs.io/projects/training/en/latest/dev/testing.html) suite locally, and they pass
-   [x] I have tested the changes on a single GPU
-   [x] I have tested the changes on multiple GPUs / multi-node setups
-   [ ] I have run the Benchmark Profiler against the old version of the code
-   [ ] If the new feature introduces modifications at the config level, I have made sure to update Pydantic Schemas and default configs accordingly

### Dependencies

-   [x] I have ensured that the code is still pip-installable after the changes and runs
-   [x] I have tested that new dependencies themselves are pip-installable.
-   [x] I have not introduced new dependencies in the inference portion of the pipeline

### Documentation

-   [x] My code follows the style guidelines of this project
-   [ ] I have updated the documentation and docstrings to reflect the changes
-   [x] I have added comments to my code, particularly in hard-to-understand areas

## Additional Notes

----
📚 Documentation preview 📚: https://anemoi-training--148.org.readthedocs.build/en/148/

----
📚 Documentation preview 📚: https://anemoi-graphs--148.org.readthedocs.build/en/148/

----
📚 Documentation preview 📚: https://anemoi-models--148.org.readthedocs.build/en/148/

### CODE REVIEW

**Comment by mchantry** [comment]
Exciting contribution, thanks.

File: training/src/anemoi/training/losses/spatial.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    power_spectra_fake = torch.abs(torch.fft.rfft2(fake_output)) ** 2
+    ratio = (power_spectra_real + epsilon) / (power_spectra_fake + epsilon)
+
+    def log10(x: torch.Tensor) -> torch.Tensor:
+        return torch.log(x) / torch.log(torch.tensor(10.0, device=x.device, dtype=x.dtype))
+
+    return (10 * log10(ratio)) ** 2
+
+
+class LogSpectralDistance(FunctionalWeightedLoss):
```
Comment (mchantry) [inline training/src/anemoi/training/losses/spatial.py]: I wonder if a different name might be better, more precise that this uses 2DFFTs for the spectra.
It's possible we may want to add a Spherical harmonic spectral distance later, and both could be interpreted as a spectral distance.

Comment (OpheliaMiralles) [reply training/src/anemoi/training/losses/spatial.py]: Which name would you suggest? I am also wondering whether this current implementation is correct (that is why I left it in draft mode) because anemoi-datasets uses flattened spatial coordinates. What I had in mind is computing the 2D fourier transform on time/batch sequences of 2D fields. I have another implementation, however, it takes as input of the loss a field shape and reshape the tensor before computing the loss, e.g. it would not work with unstructured grids. So I don't know whether to drop it or update the PR with the 2D version and if the x/y dim are not provided, it would raise an error anyway.

**Comment by mchantry** [comment]
Sorry for the slow re-review Ophelia.
I think this will work for rectilinear grids. Do you have a thought on how we could ensure that it is only used in this context, and not misused?
Would you be able to provide an update to the docs please.

**Comment by OpheliaMiralles** [comment]
> Sorry for the slow re-review Ophelia. I think this will work for rectilinear grids. Do you have a thought on how we could ensure that it is only used in this context, and not misused? Would you be able to provide an update to the docs please.

No worries, I was waiting for a signal to start working on this again. I might have some time next week or the following.

**Comment by OpheliaMiralles** [comment]
> Sorry for the slow re-review Ophelia. I think this will work for rectilinear grids. Do you have a thought on how we could ensure that it is only used in this context, and not misused? Would you be able to provide an update to the docs please.

Updated the doc and added a warning

File: training/src/anemoi/training/losses/spatial.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+) -> torch.Tensor:
+    return power_spectra_real * power_spectra_fake.conj() + power_spectra_fake * power_spectra_real.conj()
+
+
+def get_spectra(
+    fake_output: torch.Tensor,
+    real_output: torch.Tensor,
+    dims: tuple[int, int],
+) -> tuple[torch.Tensor, torch.Tensor]:
+    assert dims[0] * dims[1] == real_output.shape[2]
```
Comment (HCookie) [inline training/src/anemoi/training/losses/spatial.py]: Can you add a message to this assert, so if it is triggered, a useful message is shown

File: training/src/anemoi/training/losses/spatial.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        return torch.log(x) / torch.log(torch.tensor(10.0, device=x.device, dtype=x.dtype))
+
+    log_10 = (10 * log10(ratio)) ** 2
+    return torch.sqrt(torch.mean(log_10)) / 10
+
+
+class LogFFT2Distance(FunctionalLoss):
+    r"""The log spectral distance is used to compute the difference between spectra of two fields.
+
+    It is also called log spectral distorsion.
```
Comment (HCookie) [inline training/src/anemoi/training/losses/spatial.py]: `Distortion`?

File: training/src/anemoi/training/losses/spatial.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    dims: tuple[int, int],
+) -> tuple[torch.Tensor, torch.Tensor]:
+    assert dims[0] * dims[1] == real_output.shape[2]
+    dims_total = (*real_output.shape[:2], *dims, real_output.shape[-1])
+    power_spectra_real = torch.fft.rfft2(real_output.reshape(dims_total), dim=(-2, -3))
+    power_spectra_fake = torch.fft.rfft2(fake_output.reshape(dims_total), dim=(-2, -3))
+    return power_spectra_real, power_spectra_fake
+
+
+def log_rfft2_distance(fake_output: torch.Tensor, real_output: torch.Tensor, dims: tuple[int, int]) -> torch.Tensor:
```
Comment (HCookie) [inline training/src/anemoi/training/losses/spatial.py]: Why is the prediction being called `fake_output`?

File: training/src/anemoi/training/losses/spatial.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    power_spectra_real, power_spectra_fake = get_spectra(fake_output, real_output, dims)
+    epsilon = torch.finfo(real_output.dtype).eps  # Small epsilon to avoid division by zero
```
Comment (HCookie) [inline training/src/anemoi/training/losses/spatial.py]: Please add a docstring to this function

File: training/src/anemoi/training/losses/spatial.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        y_dim: int,
+        ignore_nans: bool = False,
+    ) -> None:
+        super().__init__(ignore_nans)
+        LOGGER.warning(
+            "LogFFT2Distance can only be used with data on 2D grids.",
+        )
+        self.x_dim = x_dim
+        self.y_dim = y_dim
+        LOGGER.warning("Please use x_dim and y_dim such that field_shape=(x_dim, y_dim).")
```
Comment (HCookie) [inline training/src/anemoi/training/losses/spatial.py]: Rather then raise a warning here, can you add a check and exception in the `calculate` function?

File: training/src/anemoi/training/losses/spatial.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        self,
+        pred: torch.Tensor,
+        target: torch.Tensor,
+        squash: bool = True,
+        *,
+        scalar_indices: tuple[int, ...] | None = None,
+        without_scalars: list[str] | list[int] | None = None,
+    ) -> torch.Tensor:
+        result = super().forward(pred, target, squash, scalar_indices=scalar_indices, without_scalars=without_scalars)
+        return torch.sqrt(torch.mean(result))
```
Comment (HCookie) [inline training/src/anemoi/training/losses/spatial.py:104]: In the case of `squash==True`, I believe the dimensions will have already been reduced and averaged. Does the `sqrt(` belong in the `calculate_difference`?

File: training/src/anemoi/training/losses/spatial.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        y_dim: int,
+        ignore_nans: bool = False,
+    ) -> None:
+        super().__init__(ignore_nans)
+        LOGGER.warning(
+            "Fourier Correlation loss can only be used with data on 2D grids.",
+        )
+        self.x_dim = x_dim
+        self.y_dim = y_dim
+        LOGGER.warning("Please use x_dim and y_dim such that field_shape=(x_dim, y_dim).")
```
Comment (HCookie) [inline training/src/anemoi/training/losses/spatial.py]: Same note as above

File: training/src/anemoi/training/losses/spatial.py (hunk 7 above)
Comment (OpheliaMiralles) [reply training/src/anemoi/training/losses/spatial.py:104]: no indeed

File: training/docs/modules/losses.rst [hunk 9]
```diff
@@ -67,6 +67,25 @@ deterministic:
       _target_: anemoi.training.losses.kcrps.KernelCRPSLoss
       # loss function kwargs here
 
+************************
+ Spatial Loss Functions
+************************
```
Comment (mchantry) [inline training/docs/modules/losses.rst:72]: Could you please specify that this only works for 2D regular grids please?

Comment (mchantry) [reply training/docs/modules/losses.rst:72]: Really nice to see documentation

Comment (OpheliaMiralles) [reply training/docs/modules/losses.rst:72]: Done!

Comment (mchantry) [reply training/docs/modules/losses.rst:72]: Sorry to be a pedant, but I think the word regular is really important, all the grids are 2D (horizontal) but being able to write it as m x n is the key component, which I think should be described as "regular".
Maybe you could even say 
** to be used only with regular 2D fields (i.e. fields that can be written as `[n_lat, n_lon]`)**

What do you think?

Comment (OpheliaMiralles) [reply training/docs/modules/losses.rst:72]: Let's hope it is good now!

File: training/src/anemoi/training/losses/loss.py [hunk 10]
```diff
@@ -68,6 +68,10 @@ def get_loss_function(
     """
     loss_config = OmegaConf.to_container(config, resolve=True)
     scalers_to_include = loss_config.pop("scalers", [])
+    # related to spatial losses
+    if config.get("x_dim", None) is not None and config.get("y_dim", None) is not None:
+        kwargs["x_dim"] = config.x_dim
+        kwargs["y_dim"] = config.y_dim
```
Comment (HCookie) [inline training/src/anemoi/training/losses/loss.py]: Why can't the `x` and `y` dims be provided in the config of the loss function rather than then in the top level config?

File: training/src/anemoi/training/losses/spatial.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+def get_spectra(
+    predicted_output: torch.Tensor,
+    real_output: torch.Tensor,
+    dims: tuple[int, int],
+) -> tuple[torch.Tensor, torch.Tensor]:
+    assert dims[0] * dims[1] == real_output.shape[2], (
+        "The product of dims must match the spatial dims of the output."
+        "Please use x_dim and y_dim such that field_shape=(x_dim, y_dim)."
+    )
+    dims_total = (*real_output.shape[:2], *dims, real_output.shape[-1])
```
Comment (HCookie) [inline training/src/anemoi/training/losses/spatial.py]: Please use the `TensorDim` Enum when referencing axis points in the tensors

File: training/src/anemoi/training/losses/loss.py [hunk 12]
```diff
@@ -69,7 +69,7 @@ def get_loss_function(
     loss_config = OmegaConf.to_container(config, resolve=True)
     scalers_to_include = loss_config.pop("scalers", [])
     # related to spatial losses
-    if config.get("x_dim", None) is not None and config.get("y_dim", None) is not None:
+    if loss_config.get("x_dim", None) is not None and loss_config.get("y_dim", None) is not None:
         kwargs["x_dim"] = config.x_dim
         kwargs["y_dim"] = config.y_dim
```
Comment (HCookie) [inline training/src/anemoi/training/losses/loss.py]: If they are in the loss config, why pull them out into the kwargs, they are used to instantiate?

Comment (OpheliaMiralles) [reply training/src/anemoi/training/losses/loss.py]: then I guess we don't need it at all indeed, could you please remove those lines? not working this afternoon but would love to see this merged

Comment (HCookie) [reply training/src/anemoi/training/losses/loss.py]: I'm unable to test this, but I have removed it, so once it's tested we can merge

File: training/src/anemoi/training/losses/spatial.py [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+import torch
+import torch.fft
+
+from anemoi.training.losses.base import FunctionalLoss
+from anemoi.training.utils.enums import TensorDim
+
+LOGGER = logging.getLogger(__name__)
+
+
+def amplitude(spectrum: torch.Tensor) -> torch.Tensor:
```
Comment (floriankrb) [inline training/src/anemoi/training/losses/spatial.py:24]: As I understand, all functions in here are private ? And only the classes inheriting from `FunctionalLoss` are public?
If so, they could be renamed `amplitude` -> `_amplitude` for instance.
